File: app/src/main/java/com/reffum/myyoutube/MainActivity.java
package com.reffum.myyoutube;

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MotionEvent;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;
import android.widget.MediaController;
import android.widget.ProgressBar;
import android.widget.SearchView;
import android.widget.Toast;
import android.widget.VideoView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.http.GenericUrl;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.gson.GsonFactory;
import com.google.api.services.youtube.YouTube;
import com.google.api.services.youtube.model.ResourceId;
import com.google.api.services.youtube.model.SearchListResponse;
import com.google.api.services.youtube.model.SearchResult;
import com.google.api.services.youtube.model.SearchResultSnippet;
import com.google.api.services.youtube.model.ThumbnailDetails;
import com.yausername.youtubedl_android.YoutubeDL;
import com.yausername.youtubedl_android.YoutubeDLException;
import com.yausername.youtubedl_android.YoutubeDLRequest;
import com.yausername.youtubedl_android.mapper.VideoInfo;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.security.GeneralSecurityException;
import java.util.ArrayList;
import java.util.Arrays;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity  {

    private static final String APPLICATION_NAME = "myyoutube";
    private final String TAG = "MyYoutube";
    private final String API_KEY = "AIzaSyBGMvsvE5t8D8p213pxuNglIQEfO--1wXU";

    YouTube mYouTube;

    private SearchRecycleViewAdapter searchRecycleViewAdapter;

    // Activity views
    private SurfaceView surfaceView;
    private SurfaceHolder surfaceHolder;
    private ProgressBar progressBar;
    private MediaController mediaController;
    private MediaPlayer mediaPlayer;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main_activity);

        // Youtube service, provide YouTube Data API v3.
        // The YouTube Data API v3 is an API that provides access to YouTube data,
        // such as videos, playlists, and channels.
        try {
            mYouTube = getService();
        } catch (GeneralSecurityException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

        searchRecycleViewAdapter = new SearchRecycleViewAdapter();
        RecyclerView video_list_recycle_view = findViewById(R.id.video_list_recycler_view);
        video_list_recycle_view.setLayoutManager(new LinearLayoutManager(this));
        video_list_recycle_view.setAdapter(searchRecycleViewAdapter);

        initViews();
        initListeners();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

        // User initial search
        if(Intent.ACTION_SEARCH.equals(intent.getAction())){
            String querry = intent.getStringExtra(SearchManager.QUERY);

            progressBar.setVisibility(View.VISIBLE);
            new Thread(() -> {
                try {
                    YouTube.Search.List request = mYouTube.search().list(Arrays.asList("id,snippet"));
                    SearchListResponse response = request.setQ(querry)
                            .setType(Arrays.asList("video"))
                            .setMaxResults(50L)
                            .execute();

                    ArrayList<VideoData> videoList = new ArrayList<>();

                    for (SearchResult searchResult : response.getItems()) {
                        ResourceId id = searchResult.getId();
                        SearchResultSnippet snippet = searchResult.getSnippet();
                        ThumbnailDetails thumbnailDetails = snippet.getThumbnails();

                        // Get video id, title and image
                        String videoId = id.getVideoId();
                        String title = snippet.getTitle();
                        String detail = snippet.getDescription();
                        String imageUrl = thumbnailDetails.getDefault().getUrl();

                        // Load image
                        URL url = new URL(imageUrl);
                        Bitmap bmp = BitmapFactory.decodeStream(url.
                                openConnection().
                                getInputStream());

                        videoList.add(new VideoData(title, detail, videoId, bmp));
                    }

                    runOnUiThread( () -> {
                       searchRecycleViewAdapter.setVideoList(videoList);
                    });
                } catch (IOException e) {
                    e.printStackTrace();
                } finally {
                    runOnUiThread( () -> {
                        progressBar.setVisibility(View.GONE);
                    });
                }
            }).start();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the options menu from XML
        MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.main_menu, menu);

        // Get the SearchView and set searchable configuration
        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        SearchView searchView = (SearchView) menu.findItem(R.id.menu_search).getActionView();
        searchView.setSearchableInfo(
                searchManager.getSearchableInfo(getComponentName())
        );
        searchView.setIconifiedByDefault(false);

        return true;
    }

    @Override
    protected void onDestroy() {
        if(mediaPlayer != null) {
            mediaPlayer.release();
            mediaPlayer = null;
        }
        super.onDestroy();
    }

    private void initViews() {
        surfaceView = findViewById(R.id.surface_view);
        surfaceHolder = surfaceView.getHolder();

        progressBar = findViewById(R.id.progress_bar);

        mediaController = new MediaController(this);
        mediaController.setAnchorView(surfaceView);
    }

    private void initListeners() {
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        mediaController.show();
        return false;
    }

    /**
     * Load youtube video in WebView
     * @param videoId
     */
    public void loadVideo(String videoId) {
        // Play youtube video
        String url = "http://www.youtube.com/watch?v=" + videoId;

        progressBar.setVisibility(View.VISIBLE);
        Disposable disposable = Observable.fromCallable(() -> {
            YoutubeDLRequest request = new YoutubeDLRequest(url);
            request.addOption("-f", "best");
            return YoutubeDL.getInstance().getInfo(request);
        })
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(streamInfo -> {
                    progressBar.setVisibility(View.GONE);
                    String videoUrl = streamInfo.getUrl();
                    if(videoUrl.isEmpty()) {
                        Toast.makeText(getApplicationContext(), "failed to get stream url",
                                Toast.LENGTH_SHORT).show();
                    } else {
                        setupVideoView(videoUrl);
                    }
                }, e -> {
                    progressBar.setVisibility(View.GONE);
                    if(BuildConfig.DEBUG)
                        Log.d(TAG, "failed to get stream url", e);
                    Toast.makeText(getApplicationContext(), "streaming failed.",
                            Toast.LENGTH_SHORT).show();
                });

    }

    private void setupVideoView(String videoUrl) throws IOException {
        mediaPlayer = new MediaPlayer();
        mediaPlayer.setSurface(surfaceHolder.getSurface());
        mediaPlayer.setDataSource(videoUrl);
        mediaPlayer.prepare();

        adjustSurfaceViewSize();

        mediaPlayer.start();

        mediaController.setMediaPlayer(new MediaController.MediaPlayerControl() {
            @Override
            public void start() {
                mediaPlayer.start();
            }

            @Override
            public void pause() {
                mediaPlayer.pause();
            }

            @Override
            public int getDuration() {
                return 0;
            }

            @Override
            public int getCurrentPosition() {
                return 0;
            }

            @Override
            public void seekTo(int i) {

            }

            @Override
            public boolean isPlaying() {
                return false;
            }

            @Override
            public int getBufferPercentage() {
                return 0;
            }

            @Override
            public boolean canPause() {
                return false;
            }

            @Override
            public boolean canSeekBackward() {
                return false;
            }

            @Override
            public boolean canSeekForward() {
                return false;
            }

            @Override
            public int getAudioSessionId() {
                return 0;
            }
        });

        mediaController.show();
    }

    void adjustSurfaceViewSize() {
        int width = surfaceView.getWidth();
        int height = surfaceView.getHeight();
        float boxWidth = width;
        float boxHeight = height;

        float videoWidth = mediaPlayer.getVideoWidth();
        float videoHeight = mediaPlayer.getVideoHeight();

        Log.i(TAG, String.format("startVideoPlayback @ %d - video %dx%d - box %dx%d", 0, (int) videoWidth, (int) videoHeight, width, height));

        float wr = boxWidth / videoWidth;
        float hr = boxHeight / videoHeight;
        float ar = videoWidth / videoHeight;

        if (wr > hr)
            width = (int) (boxHeight * ar);
        else
            height = (int) (boxWidth / ar);

        Log.i(TAG, String.format("Scaled to %dx%d", width, height));

        surfaceHolder.setFixedSize(width, height);
        mediaPlayer.seekTo(0);
    }

    public YouTube getService() throws GeneralSecurityException, IOException {
        final NetHttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
        return new YouTube.Builder(httpTransport, new GsonFactory(), request -> {
            // Add Google API_KEY in request
            GenericUrl url = request.getUrl();
            url.set("key", API_KEY);
            Log.d(TAG, "URL: " + url.toString());
        }).setApplicationName(APPLICATION_NAME)
          .build();
    }
}

File: app/src/main/java/com/reffum/myyoutube/view/MainActivity.kt
package com.reffum.myyoutube.view

import android.accounts.Account
import android.accounts.AccountManager
import android.accounts.AccountManagerFuture
import android.app.SearchManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.util.Log
import android.view.*
import android.widget.*
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.reffum.myyoutube.MediaPlaybackService
import com.reffum.myyoutube.R
import com.reffum.myyoutube.model.SearchList
import com.reffum.myyoutube.model.VideoData
import com.reffum.myyoutube.viewmodel.MainActivityViewModel
import com.reffum.myyoutube.viewmodel.SearchListAdapter
import kotlinx.coroutines.launch


class MainActivity : AppCompatActivity(),
    SearchListAdapter.ItemClickedListener,
    View.OnClickListener
{
    companion object {
        private const val TAG = "MainActivity"
    }

    private val model : MainActivityViewModel by viewModels()

    // Activity views
    private lateinit var surfaceView: SurfaceView
    private lateinit var videoTitle : TextView
    private lateinit var accountButton : Button
    private lateinit var surfaceHolder: SurfaceHolder
    private lateinit var progressBar: ProgressBar
    private lateinit var videoListRecycleView : RecyclerView
    private lateinit var mediaControllerWidget : MediaController
    private val searchRecycleViewAdapter: SearchListAdapter =
        SearchListAdapter(this)

    // Our connection to the Media Service
    private val serviceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            if(service is MediaPlaybackService.MediaServiceBinder) {
                model.mediaService = service.getService()
            }
            model.mediaService!!.setSurfaceHolder(surfaceView)
            mediaControllerWidget.setMediaPlayer(
                model.mediaService!!.getMediaControllerCallback()
            )
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            Log.d(TAG, "onServiceDisconnected called. Service crashed.")
            model.mediaService = null
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_activity)

        initViews()

        // Start media service
        startForegroundService(Intent(this, MediaPlaybackService::class.java))
        bindService(
            Intent(this, MediaPlaybackService::class.java),
            serviceConnection,
            Context.BIND_AUTO_CREATE
        )

        Log.d(TAG, "onCreate()")
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        mediaControllerWidget.show()
        return false
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)

        // Get the SearchView and set searchable configuration
        val searchManager : SearchManager = getSystemService(Context.SEARCH_SERVICE)
                as SearchManager

        val searchView = menu.findItem(R.id.menu_search).actionView as SearchView
        searchView.setSearchableInfo(
            searchManager.getSearchableInfo(componentName)
        )
        searchView.isIconifiedByDefault = false
        return true
    }

    /**
     * Handle click on video list in RecycleView
     * @param videoData
     */
    override fun onVideoItemClick(videoData: VideoData) {
        videoTitle.text = videoData.title
        SearchList.current = videoData
        loadVideo(videoData.id)
    }

    /**
     * Handle only search intent. Toolbar search create it.
     */
    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)

        if(intent == null)
            return

        // User initial search
        if(Intent.ACTION_SEARCH == intent.action) {
            lifecycleScope.launch {
                val searchString = intent.getStringExtra(SearchManager.QUERY)!!
                progressBar.visibility = View.VISIBLE
                model.searchYoutubeVideo(searchString)
                progressBar.visibility = View.GONE
            }
        }
    }

    override fun onClick(v: View?) {
        when(v?.id) {
            R.id.accounts_button -> {
                Log.d(TAG, "onClick called")
                requestOauth2()
            }
        }
    }

    /**
     * Load youtube video by its id and start play it in MediaPlayer
     * @param videoId Youtube video id
     */
    private fun loadVideo(videoId : String) {
        lifecycleScope.launch {
            // Show progress bar while video download
            progressBar.visibility = View.VISIBLE

            val directUrl: String = model.getYoutubeDirectVideoUrl(videoId)

            if(directUrl.isNotEmpty()) {
                model.playVideo(directUrl)
            } else {
                Toast.makeText(
                    applicationContext,
                    "Video load error",
                    Toast.LENGTH_SHORT
                ).show()
                Log.e(TAG, "Get stream url for $videoId error")
            }
            progressBar.visibility = View.GONE
        }
    }

    private fun initViews() {
        videoListRecycleView = findViewById<RecyclerView>(R.id.video_list_recycler_view)!!
            .apply {
                layoutManager = LinearLayoutManager(this@MainActivity)
                adapter = searchRecycleViewAdapter
            }

        // Observer for SearchList data
        val searchListObserver = Observer<List<VideoData>> {
            searchRecycleViewAdapter.notifyDataSetChanged()
        }

        SearchList.list.observe(this, searchListObserver)

        surfaceView = findViewById(R.id.surface_view)
        videoTitle = findViewById(R.id.video_title)
        surfaceHolder = surfaceView.holder
        progressBar = findViewById(R.id.progress_bar)
        mediaControllerWidget = MediaController(this)
        mediaControllerWidget.setAnchorView(surfaceView)

        accountButton = findViewById(R.id.accounts_button)
        accountButton.setOnClickListener(this)
    }

    private fun requestGoogleAccount(): Account {
        val am = AccountManager.get(this)
        val accounts = am.getAccountsByType("com.android.email")

        // System accounts list can't be empty. It must have at least 1
        assert(accounts.isNotEmpty())

        return accounts[0]
    }

    private fun requestOauth2() {
        val googleAccount = requestGoogleAccount()

        Log.d(TAG, "Google account: ${googleAccount.name}")

        val am : AccountManager = AccountManager.get(this)
        val options = Bundle()

        Log.d(TAG, "Request OAuth2")
        val result = am.getAuthToken(
            googleAccount,
            "Manage your task",
            options,
            this,

            {result : AccountManagerFuture<Bundle> ->
                Log.d(TAG, "Account success")
//                val bundle = result.result
//
//                // The token is a named value in the bundle. The name of the value
//                // is stored in the constant AccountManager.KEY_AUTHTOKEN.
//                val token = bundle.getString(AccountManager.KEY_AUTHTOKEN)

//                Log.d(TAG, "TOKEN = $token")
            },

            Handler {
                Log.d(TAG, "Account error")
                false
            }
        )

        Log.d(TAG, "Account done: ${result.isDone}")
        Log.d(TAG, "Account canceled: ${result.isCancelled}")

    }
}
